userspace/libsinsp/sinsp.cpp

```diff
@@ -1284,11 +1268,52 @@ int32_t sinsp::next(OUT sinsp_evt **puevt)
 			evt->m_dump_flags = m_replay_scap_flags;
 			m_replay_scap_evt = NULL;
 		}
-		else 
+		else
 		{
-			// If no last event was saved, invoke
-			// the actual scap_next
-			res = scap_next(m_h, &(evt->m_pevt), &(evt->m_cpuid), &(evt->m_dump_flags));
+			res = SCAP_SUCCESS;
+
+			//check if we don't have a delayed scap_evt
+			if(m_delayed_scap_evt.empty())
+			{
+				// if no pending state events continue without saving scap_evt
+				if(m_pending_state_evts.empty())
+				{
+					res = scap_next(m_h, &(evt->m_pevt), &(evt->m_cpuid), &(evt->m_dump_flags));
+				}
+				else
+				{
+					// poll+save scap_evt to compare against the state queue
+					res = m_delayed_scap_evt.next(m_h);
+				}
+			}
+
+			if(res == SCAP_SUCCESS && !m_delayed_scap_evt.empty())
+			{
+				// ts compare predicate
+				// updates state event ts if it's == -1
+				const auto check_ts = [this](const sinsp_evt* evt)
```

Comment:
I don't really love the repetition of the "weird" code in `check_ts` and `state_evts_less` :/ perhaps a macro to unify them? 
Imagine if the meaning of `-2` becomes peculiar too, we need to remember to manage it in 2 places.